### PR TITLE
Added Kafka config to support acknowledgments and MSK arn

### PR DESCRIPTION
**Pull request by kkondaka**
### Description
Added Kafka config to support acknowledgments and MSK arn
1. Added support for end-to-end acknowledgements.
    * Added configuration for enabling acknowledgements and timeout
    * Added TopicPartition commit tracker to address acknowledgements being received in different order of offsets and commit only if all the acks for all the previous offsets is received.
2. Added config for MSK ARN
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ X] New functionality includes testing.
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [X ] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java [hunk 1]
```diff
@@ -36,6 +38,23 @@ public class KafkaSourceConfig {
     @JsonProperty("authentication")
     private AuthConfig authConfig;
 
+    @JsonProperty("aws")
+    private AwsConfig awsConfig;
+
+    @JsonProperty("acknowledgments")
+    private Boolean acknowledgementsEnabled = false;
+
+    @JsonProperty("acknowledgments_timeout")
```
Comment (graytaylor0) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java:47]: While sqs does only have `acknowledgments` at the top level, I think this is a good time to make this a common config and to nest it a bit (we may need more configuration options for acknowledgments in the future)

```
acknowledgments:
  enabled: true
  timeout: "100s"
```

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaSourceCustomConsumer.java [hunk 2]
```diff
@@ -93,58 +181,59 @@ public void run() {
...
-                }
+                commitOffsets();
             }
         } catch (Exception exp) {
             LOG.error("Error while reading the records from the topic...", exp);
         }
     }
 
     private <T> Record<Event> getRecord(ConsumerRecord<String, T> consumerRecord) {
+        final String DEFAULT_KEY = "message";
```
Comment (graytaylor0) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaSourceCustomConsumer.java:192]: Nit: Could just make this a private static variable in the class

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaSourceCustomConsumer.java [hunk 3]
```diff
@@ -93,58 +181,59 @@ public void run() {
...
         }
+        data.put(key, value);
         event = JacksonLog.builder().withData(data).build();
         return new Record<Event>(event);
     }
 
-    private <T> Map<TopicPartition, OffsetAndMetadata> iterateRecordPartitions(ConsumerRecords<String, T> records) throws Exception {
-        Map<TopicPartition, OffsetAndMetadata> offsets = new HashMap<>();
+    private <T> void iterateRecordPartitions(ConsumerRecords<String, T> records, final AcknowledgementSet acknowledgementSet, Map<TopicPartition, Range<Long>> offsets) throws Exception {
+        //Map<TopicPartition, Range<Long>> offsets = new HashMap<>();
```
Comment (graytaylor0) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaSourceCustomConsumer.java:218]: Should remove this comment

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaSourceCustomConsumer.java [hunk 4]
```diff
@@ -93,58 +181,59 @@ public void run() {
...
                 }
             }
             long lastOffset = partitionRecords.get(partitionRecords.size() - 1).offset();
-            offsets.put(topicPartition, new OffsetAndMetadata(lastOffset + 1));
+            long firstOffset = partitionRecords.get(0).offset();
+            Range<Long> offsetRange = Range.between(firstOffset, lastOffset);
+            offsets.put(topicPartition, offsetRange);
         }
-        return offsets;
+        //return offsets;
```
Comment (graytaylor0) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/KafkaSourceCustomConsumer.java:236]: And remove comment here

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/TopicPartitionCommitTracker.java [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+            if (merged) {
+                offsetMinMap.remove(min);   
+                offsetMaxMap.remove(max);   
+            }
+            max = entry.getMaximum();
+            offsetMaxMap.remove(max);
+            Range<Long> newEntry = Range.between(min, max);
+            offsetMaxMap.put(max, newEntry);
+            offsetMinMap.put(min, newEntry);
+            merged = true;
```
Comment (graytaylor0) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/consumer/TopicPartitionCommitTracker.java:60]: This looks a bit duplicated with the previous block. Is it possible to run this into a reusable function?

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java [hunk 6]
```diff
@@ -11,13 +11,15 @@
...
 import java.util.List;
+import java.time.Duration;
 
 /**
  * * A helper class that helps to read user configuration values from
  * pipelines.yaml
  */
 
 public class KafkaSourceConfig {
+    static final Duration DEFAULT_ACKNOWLEDGEMENTS_TIMEOUT = Duration.ofSeconds(5);
```
[the remark this answers is not in the document]
Comment (kkondaka) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java:22]: ok.

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java (hunk 1 above)
Comment (graytaylor0) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/KafkaSourceConfig.java:47]: There is a consideration to only have one setting which will also enable acknowledgments

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/AwsConfig.java [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+
+package org.opensearch.dataprepper.plugins.kafka.configuration;
+
+import com.fasterxml.jackson.annotation.JsonProperty;
+import jakarta.validation.constraints.Size;
+
+public class AwsConfig {
+    @JsonProperty("msk_arn")
+    @Size(min = 20, max = 2048, message = "mskArn length should be between 20 and 2048 characters")
+    private String awsMskArn;
```
[the remark this answers is not in the document]
Comment (graytaylor0) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/configuration/AwsConfig.java:14]: How do you get permissions to read from aws kafka then?

**Review by graytaylor0 — APPROVED** [review]

**Review by asifsmohammed — APPROVED** [review]

**Pull request merged**